core/src/main/java/kafka/server/handlers/DescribeTopicPartitionsRequestHandler.java

```diff
@@ -0,0 +1,137 @@
+        AuthHelper authHelper,
+        KafkaConfig config
+    ) {
+        this.metadataCache = metadataCache;
+        this.authHelper = authHelper;
+        this.config = config;
+    }
+
+    public DescribeTopicPartitionsResponseData handleDescribeTopicPartitionsRequest(RequestChannel.Request abstractRequest) {
+        if (metadataCache instanceof ZkMetadataCache) {
+            throw new InvalidRequestException("ZK cluster does not handle DescribeTopicPartitions request");
+        }
+        KRaftMetadataCache kRaftMetadataCache = (KRaftMetadataCache) metadataCache;
+
+        DescribeTopicPartitionsRequestData request = ((DescribeTopicPartitionsRequest) abstractRequest.loggableRequest()).data();
+        Set<String> topics = new HashSet<>();
+        boolean fetchAllTopics = request.topics().isEmpty();
+        DescribeTopicPartitionsRequestData.Cursor cursor = request.cursor();
+        if (fetchAllTopics) {
+            if (cursor != null) {
+                // Includes the cursor topic in case the cursor topic does not exist anymore.
+                topics.add(cursor.topicName());
+                kRaftMetadataCache.getAllTopicsAfterTopic(cursor.topicName()).foreach(topic -> topics.add(topic));
+            } else {
+                kRaftMetadataCache.getAllTopics().foreach(topic -> topics.add(topic));
+            }
+        } else {
+            request.topics().forEach(topic -> {
+                String topicName = topic.name();
+                if (cursor == null || topicName.compareTo(cursor.topicName()) >= 0) {
+                    topics.add(topic.name());
+                }
+            });
+
+            if (cursor != null && !topics.contains(cursor.topicName())) {
+                // The topic in cursor must be included in the topic list if provided.
+                throw new InvalidRequestException("DescribeTopicPartitionsRequest topic list should contain the cursor topic: " + cursor.topicName());
+            }
+        }
+
+        // Do not disclose the existence of topics unauthorized for Describe, so we've not even checked if they exist or not
+        Set<DescribeTopicPartitionsResponseTopic> unauthorizedForDescribeTopicMetadata = new HashSet<>();
+
+        Iterator<String> authorizedTopics = topics.stream().filter(topicName -> {
+            boolean isAuthorized = authHelper.authorize(
+                abstractRequest.context(), DESCRIBE, TOPIC, topicName, true, true, 1);
+            if (!fetchAllTopics && !isAuthorized) {
+                // We should not return topicId when on unauthorized error, so we return zero uuid.
+                unauthorizedForDescribeTopicMetadata.add(describeTopicPartitionsResponseTopic(
+                    Errors.TOPIC_AUTHORIZATION_FAILED, topicName, Uuid.ZERO_UUID, false, Collections.emptyList())
+                );
+            }
+            return isAuthorized;
+        }).sorted().iterator();
+
+        int firstPartitionId = cursor == null ? 0 : cursor.partitionIndex();
```

Comment:
If cursor topic is not authorized, we'd skip partitions in the next authorized topic.

Reply:
Done.